### ISSUE TITLE
No clean way to ignore differing versions of swank

### ISSUE THREAD

**Issue by erjoalgo** [slime/slime#250]
There is currently no clean way of ignoring the annoying "versions differ" warning, because the comparison against the 'ignore symbol is done on a variable that is bound from the '(swank:connection-info)' object, which the emacs user cannot modify. Additionally, the call to slime-check-version is started asyncronously from the call to 'slime-connect in the emacs user's code, so even a let binding of some flag to 'ignore wouldn't be readable by the time slime-check-version is called. 
It seems the appropriate way to do this is to add a global option to slime which the user can configure.

```lisp
(defun slime-check-version (version conn)
  (or (equal version slime-protocol-version)
      (equal slime-protocol-version 'ignore) ;;this makes no sense, as this is bound from the (swank:connection-info) object
      slime-ignore-protocol-version-p;; add a global option that the emacs user can configure
      (y-or-n-p
       (format "Versions differ: %s (slime) vs. %s (swank). Continue? "
               slime-protocol-version version))
      (slime-net-close conn)
      (top-level)))
```

**Comment by luismbo** [comment]
Doesn't setting `swank::*swank-wire-protocol-version*` to `'swank::ignore` on the Lisp side work?

**Comment by erjoalgo** [comment]
You want to be able to do ths from the emacs side also. Perhaps you have no control over the swank server.

**Comment by luismbo** [comment]
@ellerh What's your take on this? I'm uneasy about making it too easy to ignore the protocol version check. Hmm. How hard do you think it would be to maintain actual protocol versions?

**Comment by ellerh** [comment]
> the comparison against the 'ignore symbol is done on a variable that is bound from the '(swank:connection-info)' object, which the emacs user cannot modify. 

This is simply not true. `slime-protocol-version` is an ordinary global variable that is only initialized when slime.el is loaded.  Setting it interactively with `M-x set-variable slime-protocol-version` works just fine.  Setting it in .emacs with something like
`(add-hook 'slime-load-hook (lambda () (setq slime-protocol-version 'ignore)))`
also works and is relatively clean.

> How hard do you think it would be to maintain actual protocol versions?

The prospect of spending time on non-reproducable problems that occur because some users are too lazy to keep slime and swank in sync doesn't sound exciting to me.

**Comment by erjoalgo** [comment]
Well. You're right I overlooked that 'slime-protocol-version is an ordinary global variable. However, this means that the cannonical way of setting an invocation-specific setting, ie via let binding, will not work, since slime-check-version is called asyncronously and is not a descendant of 'slime-connect in the stack.

(let* (
	 (slime-protocol-version 'ignore)
      )                      
    (slime-connect "127.0.0.1" "4005"); user will be prompted if versions disagree
    )

    "because some users are too lazy to keep slime and swank in sync"

Did it occur to you that user might have to deal with multiple swank servers with different version support? Or do you just enjoy being rude to people and making idiotic assumptions?

**Comment by ellerh** [comment]
> Did it occur to you that user might have to deal with multiple swank servers with different version support? 

Yes, but that's not something that I ever wanted to support.

> Or do you just enjoy being rude to people and making idiotic assumptions?

I don't particulary like it  when complete strangers lecture me about the workings of a piece of code when they a) make statements that are factually incorrect b) call me rude for exposing their flawed thinking c) expect a great deal of respect from me while they themself don't make any effort to be friendly.

**Comment by luismbo** [comment]
@ellerh, @erjoalgo, guys settle down!

@erjoalgo I think Elmut's solution for your problem is adequate, do you agree?

**Comment by erjoalgo** [comment]
Actually, kept trying various ways to modify the setting without luck, so I gave up and resorted to the hack below. It seems that simply `(setq slime-protocol-version 'ignore)`, which I had tried before, does not work when slime has not been loaded. But I cannot find the appropriate hook where to add this change. 

The variable is defined here, 
```
(defvar slime-protocol-version nil)
(setq slime-protocol-version
      (eval-when-compile (slime-changelog-date)))
```
but it's not clear to me when in my setup slime gets compiled. So no, it's not really too ordiary a variable, and it's far from obvious how to modify this setting.

@ellerh 

    "Yes, but that's not something that I ever wanted to support."

So does this means you own slime? 

    "I don't particulary like it when complete strangers lecture me about the workings of a piece of code when they a) make statements that are factually incorrect b) call me rude for exposing their flawed thinking c) expect a great deal of respect from me while they themself don't make any effort to be friendly."

1. I called you rude for being rude
2. I wasn't being friendly or unfriendly, I was pointing out an issue in the code, when you decided to start making unecessary assumptions and throwing insults
3. As I point out in this comment, it's far from intuitive and actually still not clear to me how to change this setting, so as it stands the issue is a valid one that hopefully someone else who is a little bit more mature can address.


To avoid further waste of time, I addded this hack to the broken slime-check-version.

```
(defadvice slime-check-version (before disable-annoying-prompt activate)
  ;;
  (setq slime-protocol-version 'ignore)
  )
```

**Comment by luismbo** [comment]
![Internet argument](https://imgs.xkcd.com/comics/internet_argument.png)

**Comment by luismbo** [comment]
@erjoalgo did you try @ellerh's `slime-load-hook` suggestion?

**Comment by erjoalgo** [comment]
Actually, I was searching for slime hooks using apropos, and I did not see that hook, but I tried other hooks, like slime-compilation-finished-hook, slime-connected-hook, slime-mode-hook, and they failed.

    `(add-hook 'slime-load-hook (lambda () (setq slime-protocol-version 'ignore)))`

does work and I do think it's somewhat clean, although finding this hook in the first place was not intuitive to me. I'm not sure why I could not find that hook via apropos. I still think changing this setting is far from clear and should be documented more thoroughly. 

So, in summary. Some common, intuitive things that common people like me might try:

1. 
```(let* ( (slime-protocol-version 'ignore) ) 
(slime-connect "127.0.0.1" "4005"); does not work )
```

2. 

```
(progn 
(setq slime-protocol-version 'ignore)
(slime-connect "127.0.0.1" "4005")); does not work
```

3.

```
;;search for slime hooks
(add-hook 'slime-compilation-finished-hook (setq slime-protocol-version 'ignore))
(add-hook 'slime-connected-hook (setq slime-protocol-version 'ignore))
(add-hook 'slime-mode-hook (setq slime-protocol-version 'ignore))
```

4. Have to dig into the source, figure out how the var is defined:
```
(defvar slime-protocol-version nil)
(setq slime-protocol-version
      (eval-when-compile (slime-changelog-date)))
```

Figure out how eval-when-compile works, figure out when slime gets compiled in your setup, then somehow figure out that the hook which doesn't seem to exist is the right solution.

    `(add-hook 'slime-load-hook (lambda () (setq slime-protocol-version 'ignore)))`

The solution itself is clean but the process for finding it is too obscure. So hopefully add some documentation to address this.

**Comment by luismbo** [comment]
@erjoalgo I agree it's obscure, I didn't know about it either, but one could argue that it's a feature, not a bug. The slime/swank protocol is fairly complex if you take into account that there's a bunch of dependencies between the elisp and lisp sides. Mismatching the two is asking for trouble and leads to bogus bug reports which is unfortunate for everyone involved. So, ignoring that check is not really something we'd like to encourage.

I understand your use case, though. Not sure what the right thing to do is, sorry.

**Comment by capitaomorte** [comment]
> Not sure what the right thing to do is, sorry.

You could provide a proper customizable variable like `slime-ignore-protocol-mismatches`, that wouldn't need `slime-load-hook`. 

In any case, to support let-binding around `slime` or `slime-connect` you need to save and restore relevant bindings in `slime-timer-call` and `slime-eval-async`. 

Also, using `eval-after-load` or `with-eval-after-load` is arguably a more idiomatic way than using something like `slime-load-hook`.

**Comment by luismbo** [comment]
Right, but you're missing the point. The question is not "is it possible to conveniently do it? " but rather "do we want to promote it?".

**Comment by capitaomorte** [comment]
You dilemma seems to be about "maintaining actual protocol versions". I was just offering new undiscussed options that don't affect that decision at all, because @erjoalgo apparently doesn't care, he just needs to ignore mismatches on a case by case basis. I have no wish to address policy whatsoever.

**Comment by erjoalgo** [comment]
I agree with @capitaomorte, currently the slime-protocol-version variable is "overloaded" to hold either the actual slime protocol version or the 'ignore flag, so for example slime-check-version couldn't tell the difference between matching protocol versions and a mismatch that the user choses to ignore (and perhaps just issue a warning).

There's a need for a separate 'slime-ignore-protocol-version-p' flag that is independent of the actual protocol version variable, and its value should survive across the async call from slime-connect (I guess by passing them through slime-timer-call and slime-eval-async as capitaomorte suggests).

**Comment by luismbo** [comment]
@erjoalgo I know that you agree. Do you understand the point of view that making it easy to ignore the version check is not something necessarily desirable?

**Comment by erjoalgo** [comment]
Just like Tavora said, 

    "You dilemma seems to be about "maintaining actual protocol versions".
    I was just offering new undiscussed options that don't affect that
    decision at all, because @erjoalgo apparently doesn't care, he just
    needs to ignore mismatches on a case by case basis. I have no wish to
    address policy whatsoever. "

I'm just pointing out that the it's bad design the way it is now. If you're saying that this is done intentionally in order to make the user's life harder, my response is that I'm not interested in baby-sitting users, who are most likely developers trying to get work done. I wasted more time than necessary on this bug and I'm trying to spare future users of the same hassle.

In my case, I sometimes need to connect to a swank server by a project that uses an older version of swank. In such cases, I know that slime protocol versions will differ. I also know that for my purpose, things will be ok, and I really don't need an annoying prompt every time, especially when such connections are being made as part of other, non user-facing routines.

It seems that you and others are not interested in fixing this issue, and I'm not going to waste more time advocating for what should be a very simple fix. I just feel sorry for others who run into this issue again in the future (actually, this issue had already been opened a while back).

**Comment by attila-lendvai** [comment]
> my response is that I'm not interested in baby-sitting users, who are most likely developers trying to get work done. I wasted more time than necessary on this bug and I'm trying to spare future users of  the same hassle.

i don't want to disappoint you too much, but the slime project maintainers, which is practically Helmut for the past few years due to his respected authority, has a long history of not welcoming this otherwise commendable attitude of trying to make life easier for other users, especially newcomers. (e.g. i had a pending patch to set utf-8 as the default encoding for so many years, that by the time someone with the commit bit considered it, the entire thing was obsolete because of a rewrite... and meanwhile this unlucky default generated regular mails to the mailing list.)

i very well understand that Helmut doesn't want to waste his time on silly bug reports -- i don't want that either -- , but i don't understand the attitude of not even considering a search for win-win solutions and ignoring the time the *users* waste (read: other programmers, i.e. potential contributors). after all, if the maintainer's time is so much more valuable than the users', then why publish an opensource project to begin with?

i think the situation with the slime project maintenance is a good example of the bad attitude the common lisp community is famous for... it's one thing to be raw and direct, but completely ignoring the users' needs is another level.

as the result of this attitude i can hardly think of an IDE that are still in use, and yet it evolves slower than slime does.

FTR, in my view a raw but reasonable response could have been this: in the current project setup an increase in clueless bug reports due to mismatched slime/swank versions will consume more of my time, so i will only commit solutions that will at most only marginally increase the number of such bug reports, and i am otherwise not interested in working on this myself.

**Comment by luismbo** [comment]
I have my opinion about what went wrong in this discussion and I don't think Helmut's *in your face* style of communication is the sole culprit, but I'd rather not discuss that too much here, send me personal email if you must vent about it.

@erjoalgo, let's try again. I'm trying to take two competing requirements into account: (1) make it easy to ignore the version check when the user knows what he's doing, (2) avoid problems when too many people start ignoring the version check. Again, I know that you care about (1), not (2), but I feel that accusing me of not wanting to fix (1) just because I care about (2) is unfair.

So, for example, if instead of a presenting a prompt, SLIME simply warned about the version mismatch, would that be a reasonable solution to your problem? If so, then it would be better than having to customize some variable. What do you think?

**Comment by attila-lendvai** [comment]
FTR, i personally prefer to talk in the open about issues that directly affect many people, like the management of an opensource project.

but i got the message, no more random venting here, if for nothing else because it's off topic.

**Comment by capitaomorte** [comment]
> I'm trying to take two competing requirements into account: (1) make it easy to ignore the version check when the user knows what he's doing, (2) avoid problems when too many people start ignoring the version check. Again, I know that you care about (1), not (2), but I feel that accusing me of not wanting to fix (1) just because I care about (2) is unfair.

Hasn't it already been fully demonstrated in this thread that achieving (1) is already a perfectly supported, legal and easy thing to do? If so, your argument doesn't make much sense, does it? The problem, at least in my view, is that it is hard to achieve (3): ignore the version check on a per-`slime-connect`  basis.

**Comment by capitaomorte** [comment]
But the warning idea sounds pretty good too.

**Comment by erjoalgo** [comment]
@capitaomorte Actually, like I said in one of my previous comments, I don't even think the proposed solution to 1 is a legal and easy or intuitive thing to do. First, it is far from intuitive, as I have shown, especially with the (eval-after-compile ) definition and the fact that the variable doesn't carry over from the slime-connect context (ie let-binding is useless). But also currently there is a single variable, slime-protocol-version that has two "types" simultaneously:

1. A string informing of the slime protocol version
2. A boolean flag indicating whether the user wants to force a connection despite a potential version mismatch

These are two pieces of information that should be kept separately, for the sake of clarity and cleanness. Also there are at least three possible things that slime-check-protocol-version might want to do, for example, 

1. versions match=> connect silently
2. versions mismatch, no 'ignore flag=> annoying y-n  prompt
3. versions mismatch, yes 'ignore flag=> non-intrusive warning

This is impossible to implement right now because slime-check-protocol-version can only act in two ways.

So, like I've said before, I suggest:

1. add a slime-ignore-protocol-version-p which exclusively stores whether user wants to force a connection
2. let this variable pass through to async call context

**Comment by luismbo** [comment]
@erjoalgo did you miss my question?

**Comment by erjoalgo** [comment]
"@erjoalgo did you miss my question?"

I hope not, I think it should have been answered by my last comment, which details what I think would be a reasonable solution.

    "So, for example, if instead of a presenting a prompt, SLIME simply warned about the version mismatch, would that be a reasonable solution to your problem? If so, then it would be better than having to customize some variable. What do you think?"

Like I said, I think slime should show a non-intrusive warning about version mismatch iff 

1. Versions do not match, and
2. User has opted to ignore

By default, I have no problem with making an intrusive y-n prompt when versions differ. But definitely let the user cleanly and intuitively disable this annoyance without having to waste as much time as I had to.

I think my position is clear enough on this, I've basically been repeating the same thing over and over.

**Comment by luismbo** [comment]
@erjoalgo My question was a bit more specific: I was trying to altogether remove the need to explicitly customize a variable in order to ignore the version check. Do you have any comments on that particular idea?

**Comment by erjoalgo** [comment]
I don't have any new comments on that particular idea, since I've already described as accurately as possible what I think should be the correct behaviour: 

1. By default, user should be made aware as intrusively as possible about version mismatch
2. User has an easy, intuitive way of opting out of this intrusion, by customizing a variable, which is the normal emacs way of doing things

So, no, I don't think the user should not have to customize a variable to opt out of the intrusive version prompt.

**Comment by luismbo** [comment]
@erjoalgo I'll buy that solution. Can you send a pull request?

**Comment by erjoalgo** [comment]
Sure. Thanks

**Comment by bigos** [comment]
When you search for info about the bug, Google sends you here. Can someone append a recommended solution at the end? I've spent much time and I'm none wiser.

**Comment by luismbo** [comment]
@bigos do you want to forcefully use mismatched SLIME and SWANK or do you want to debug why you have a mismatched SLIME and SWANK?

**Comment by bigos** [comment]
Ideally, I would love to have matched slime and swank. However, it is hard to find newbie-friendly information on how to solve the problem.
If there are no problems I thought that forcefully using mismatched slime and swank would be the second option. But I have no idea what would be the consequences of using mismatched slime and swank. Where I can find info about it? I have encountered the problem several times over the years and never found a solution. Each time the problem mysteriously went away after annoying me for a period of time. When you google for the problem search engine sends you here. Ideally, someone would put in the documentation the exact text of the error so Google would send you there.

**Comment by luismbo** [comment]
@bigos open up IELM (using <kbd>M-x ielm</kbd>) and try this:

```
ELISP> (require 'slime)
slime
ELISP> slime-version
"2.23"
ELISP> (find-lisp-object-file-name 'slime 'defun)
"/path/to/slime/slime.el"
ELISP> (concat slime-path slime-backend)
"/path/to/slime/swank-loader.lisp"
```

The swank and slime directories should be the same. If they're not, start from an clean init.el file and follow the manual instructions. HTH!

**Comment by bigos** [comment]
There might be another. I saw in quicklisp folders at home the slime version I need. Perhaps tomorrow at work, when I update quicklisp the problem, will go away?

**Comment by luismbo** [comment]
If you follow my debugging instructions, we should get plenty of information to figure out a solution.

**Comment by bigos** [comment]
I have tried your instructions in Emacs repl and it seems that trying to update quicklisp might fix the problem. Before I did not know where to look for the solution. Your instructions have helped. Thank you.

**Comment by bigos** [comment]
I confirm that going to Lisp REPL and updating quicklisp has solved the problem.

    (ql:update-client)
    (ql:update-all-dists)

After restrting Emacs the problem has disappeared. The following line in my Emacs config loads file from quicklisp that loads swank. That is why I did not know where to look for solution. Well, I have learned something new.

https://github.com/bigos/prelude/blob/master/personal/organised.el#L168
